[Serialization] Always list the bridging header before any imports

jrose-apple opened the pull request:
This mirrors how a bridging header is processed when compiling source files: before any of the imports. This is important for LLDB to recreate the source environment as closely as possible to how the compiler does it; in the test case being added involving a non-modular header file, failure to do so resulted in a deserialization cross-reference crash.

Note that Serialization still sorts imports, which normal resolution of imports in source does not do. So we're still not consistent. But this is less important than handling textual includes (bridging headers) before modular imports.

rdar://problem/40471329

jrose-apple commented:
Thanks to Adrian for heroically coming up with a reduced test case!

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented on validation-test/Serialization/Inputs/bridging-header-first/AmbivalentProtocol.h:
```diff
@@ -0,0 +1,12 @@
+// #if __clang__
+//
+// #ifndef AP
+// #define AP
```
Oops, left some testing code in.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test source compatibility

adrian-prantl reviewed (commented):
Whoa, you fixed it? Awesome!

jrose-apple committed:
[Serialization] Keep buffers alive if they may have diagnostics

This can only happen in one case today: a module imports a bridging
header, but the header on disk has disappeared, and now we need to
fall back to the (often inadequate) version that's stored inside the
swiftmodule file. Even if the module fails to load, the bridging
header has already been imported, and so anything else that happens
might still emit diagnostics and need that text to be alive, which
means we need to keep the buffer alive too.

jrose-apple committed:
[Serialization] Always list the bridging header before any imports

This mirrors how a bridging header is processed when compiling source
files: before any of the imports. This is important for LLDB to
recreate the source environment as closely as possible to how the
compiler does it; in the test case being added involving a non-modular
header file, failure to do so resulted in a deserialization
cross-reference crash.

Note that Serialization still sorts imports, which normal resolution
of imports in source does not do. So we're still not consistent. But
this is less important than handling textual includes (bridging
headers) before modular imports.

rdar://problem/40471329
